### PR TITLE
LPS-186040 | Automate FeatureFlags.testcase

### PR DESCRIPTION
**Pull request by eltonccdantas**
[Ticket](https://liferay.atlassian.net/browse/LPS-186037)

image

### CODE REVIEW

**Comment by eltonccdantas** [comment]
ci:help

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
ci:stop

**Comment by eltonccdantas** [comment]
ci:stop:ci:test:peds-sm

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
ci:stop:peds-sm

**Comment by eltonccdantas** [comment]
ci:test:st

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/paths/FeatureFlag.path [hunk 1]
```diff
@@ -12,7 +12,7 @@
 <tbody>
 	<tr>
 		<td>TOGGLE_ROW</td>
-		<td>//*[@class="list-group-title"][contains(text(),'${title_row}')]//ancestor::ul//label//span[contains(@class,'toggle-switch-label')][contains(text(),'${toggle_state}')]</td>
+		<td>//li[(contains(.,'${title_row}')) or (contains(@for,'${title_row}'))]//span[text()='${toggle_state}']</td>
```
Comment (eltonccdantas) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/paths/FeatureFlag.path:15]: I updated the path to work the same way.

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 2]
```diff
@@ -1,7 +1,7 @@
 @component-name = "portal-feature-flag"
 definition {
 
-	property custom.properties = "feature.flag.LPS-162766=false${line.separator}feature.flag.LPS-162766.dependencies=LPS-162765${line.separator}feature.flag.LPS-162766.description=Create your own terms and conditions pages${line.separator}feature.flag.LPS-162766.title=Terms and Conditions pages${line.separator}feature.flag.LPS-162766.type=release";
+	property custom.properties = "feature.flag.LPS-162765=false${line.separator}feature.flag.LPS-162765.description=Dependency FF Test Description${line.separator}feature.flag.LPS-162765.title=Dependecy FF Test${line.separator}feature.flag.LPS-162765.type=release${line.separator}feature.flag.LPS-162766=false${line.separator}feature.flag.LPS-162766.dependencies=LPS-162765${line.separator}feature.flag.LPS-162766.description=Create your own terms and conditions pages${line.separator}feature.flag.LPS-162766.title=Terms and Conditions pages${line.separator}feature.flag.LPS-162766.type=release";
```
Comment (eltonccdantas) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:4]: I added the properties for the dependency.

**Comment by eltonccdantas** [comment]
ci:test:peds-sm

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
ci:stop:peds-sm

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
ci:test:peds-sm

**Comment by eltonccdantas** [comment]
ci:test:peds-sm

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by john-co** [comment]
Just started reviewing... :hourglass_flowing_sand:

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 3]
```diff
@@ -48,17 +48,17 @@ definition {
...
 				toggle_state = "Enabled");
 		}
 
 		task ("Then assert the toggle is disabled state") {
-			AssertElementPresent(
+			AssertElementNotPresent(
 				locator1 = "FeatureFlag#TOGGLE_ROW",
 				title_row = "Terms and Conditions pages",
-				toggle_state = "Disabled");
+				toggle_state = "Enabled");
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:61]: Why is this change needed?

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 4]
```diff
@@ -48,17 +48,17 @@ definition {
 		}
 
 		task ("When the administrator disables a feature flag") {
-			Check.checkToggleSwitch(
+			ClickNoError(
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:51]: Why is this needed?

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 5]
```diff
@@ -69,7 +69,7 @@ definition {
 
 		task ("And a feature flag is disabled") {
 			if (IsElementPresent(locator1 = "FeatureFlag#TOGGLE_ROW", title_row = "Terms and Conditions pages", toggle_state = "Enabled")) {
-				Check.checkToggleSwitch(
+				ClickNoError(
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:72]: Why is this change needed?

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 6]
```diff
@@ -91,6 +91,38 @@ definition {
...
 
+	@description = "LPS-167698 - This tests asserts that a feature flag can be enabled with dependencies."
+	@priority = 3
+	test CanBeEnabledWithDependencies {
+		task ("And a feature flag is disabled") {
+			if (IsElementPresent(locator1 = "FeatureFlag#TOGGLE_ROW", title_row = "Terms and Conditions pages", toggle_state = "Enabled")) {
+				ClickNoError(
+					locator1 = "FeatureFlag#TOGGLE_ROW",
+					title_row = "Terms and Conditions pages",
+					toggle_state = "Enabled");
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:102]: Why can't we use Check.checkToggleSwitch?

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 7]
```diff
@@ -91,6 +91,38 @@ definition {
...
+			}
+		}
+
+		task ("When the administrator enables a feature flag") {
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:106]: When the administrator enables a feature flag that is a dependent

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 8]
```diff
@@ -91,6 +91,38 @@ definition {
...
+			Check.checkToggleSwitch(
+				locator1 = "FeatureFlag#TOGGLE_ROW",
+				title_row = "Terms and Conditions pages",
+				toggle_state = "Disabled");
+		}
+
+		task ("Then dependencies can be enabled") {
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:113]: Then a feature flag that depends on the enabled feature flag can be enabled

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 9]
```diff
@@ -108,4 +140,22 @@ definition {
...
+	@priority = 3
+	test CannotBeEnabledByADependency {
+		task ("When the user tries to enable a dependency before enabling a FF") {
+			Check.checkToggleSwitch(
+				locator1 = "FeatureFlag#TOGGLE_ROW",
+				title_row = "Dependency FF Test",
+				toggle_state = "Disabled");
+		}
+
+		task ("Then assert a that feature flag cannot be enabled by a dependency") {
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:153]: Then assert the dependency toggle remains in a disabled state

Comment (john-co) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:153]: Also add `And Then assert the dependency toggle is not editable`

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 10]
```diff
@@ -108,4 +140,22 @@ definition {
...
+			AssertElementNotPresent(
+				locator1 = "FeatureFlag#TOGGLE_ROW",
+				title_row = "Terms and Conditions pages",
+				toggle_state = "Enabled");
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:157]: Need to assert The Dependency FF Test is not enabled. Not Terms and Conditions.

**Review by john-co — CHANGES_REQUESTED** [review]
Thanks @eltonccdantas , please see review comments.

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 11]
```diff
@@ -1,7 +1,7 @@
 @component-name = "portal-feature-flag"
 definition {
 
-	property custom.properties = "feature.flag.LPS-162766=false${line.separator}feature.flag.LPS-162766.dependencies=LPS-162765${line.separator}feature.flag.LPS-162766.description=Create your own terms and conditions pages${line.separator}feature.flag.LPS-162766.title=Terms and Conditions pages${line.separator}feature.flag.LPS-162766.type=release";
+	property custom.properties = "feature.flag.LPS-162765=false${line.separator}feature.flag.LPS-162765.description=Create your own terms and conditions pages${line.separator}feature.flag.LPS-162765.title=Terms and Conditions pages${line.separator}feature.flag.LPS-162765.type=release${line.separator}feature.flag.LPS-162766=false${line.separator}feature.flag.LPS-162766.dependencies=LPS-162765${line.separator}feature.flag.LPS-162766.description=Dependency FF Test Description${line.separator}feature.flag.LPS-162766.title=Dependency FF Test${line.separator}feature.flag.LPS-162766.type=release";
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:4]: Why not keep LPS-162766 as `Create your own terms and conditions pages`? Wouldn't this impact the other tests that depend on LPS-162766?

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 4 above)
Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:51]: I think the logic was inverted there since `When the administrator disables a feature flag` we needed to `uncheck` instead of `check`

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 5 above)
Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:72]: I think that after I changed the `xpath`, some `Check.checkToggleSwitch` or `Uncheck` stopped working, so I needed to replace it.

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 6 above)
Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:102]: That case we need to disable, the `Uncheck.uncheckToggleSwitch` does not work with that path.

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 11 above)
Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:4]: If you look at old poshi reports([#1](https://testray.liferay.com/reports/production/logs/2023-10/test-1-26/test-portal-acceptance-pullrequest(master)/11693/functional-tomcat90-mysql57-jdk8/0/1/LocalFile.FeatureFlag_CanBeDisabled/index.html.gz) ,[#2](https://testray.liferay.com/reports/production/logs/2023-10/test-1-26/test-portal-acceptance-pullrequest(master)/11693/functional-tomcat90-mysql57-jdk8/0/1/LocalFile.FeatureFlag_CanBeEnabled/index.html.gz) ), it was working as a dependency, so I decided [to refactor it](https://github.com/liferay-platform-experience/liferay-portal/pull/343#discussion_r1372228380).

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 3 above)
Comment (john-co) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:61]: if it's a matter of fitting with the task better, I would say it was better before. Since we're asserting it's Disabled state

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 11 above)
Comment (john-co) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:4]: No, it's not working as a dependency for those tests. There's no other feature flag that depends on it referenced in the tests.

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/paths/FeatureFlag.path (hunk 1 above)
Comment (john-co) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/paths/FeatureFlag.path:15]: [(contains(.,'${title_row}')) or (contains(@for,'${title_row}'))]

What's the difference between these?

Comment (john-co) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/paths/FeatureFlag.path:15]: https://testray.liferay.com/reports/production/logs/2023-10/test-1-31/test-portal-acceptance-pullrequest(master)/9759/functional-tomcat90-mysql57-chrome100-jdk8/30/14/LocalFile.FeatureFlag_CanBeEnabled/index.html.gz?authuser=0

The other tests pass with the old xpath. I would prefer the old xpath than the new one since the class attributes help describe the toggle better.

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
ci:test:sf

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
Investigating it... 🕵🏻

**Comment by eltonccdantas** [comment]
ci:test:peds-sm:nocompile

**Comment by eltonccdantas** [comment]
Hey @john-co , may we forward this one?

image

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 12]
```diff
@@ -111,6 +111,16 @@ definition {
...
+			WaitForElementPresent(
+				locator1 = "FeatureFlag#TOGGLE_ROW",
+				title_row = "Dependency FF Test",
+				toggle_state = "Disabled");
+
+			ScrollWebElementIntoView(
+				locator1 = "FeatureFlag#TOGGLE_ROW",
+				title_row = "Dependency FF Test",
+				toggle_state = "Disabled");
+
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:123]: Is this necessary? Remnants from troubleshooting?

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase [hunk 13]
```diff
@@ -144,6 +154,18 @@ definition {
...
+			if (IsElementPresent(locator1 = "FeatureFlag#TOGGLE_ROW", title_row = "Terms and Conditions pages", toggle_state = "Enabled")) {
+				ClickNoError(
+					locator1 = "FeatureFlag#TOGGLE_ROW",
+					title_row = "Terms and Conditions pages",
+					toggle_state = "Enabled");
+
+				ScrollWebElementIntoView(
+					locator1 = "FeatureFlag#TOGGLE_ROW",
+					title_row = "Dependency FF Test",
+					toggle_state = "Disabled");
```
Comment (john-co) [inline modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:166]: same comment as above

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 12 above)
Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase:123]: I have the impression you told me to add those lines to mimic the user's behavior. I'll remove them now!

Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase]: Done

File: modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase (hunk 13 above)
Comment (eltonccdantas) [reply modules/apps/feature-flag/feature-flag-test/src/testFunctional/tests/FeatureFlag.testcase]: Done

**Comment by john-co** [comment]
ci:forward
